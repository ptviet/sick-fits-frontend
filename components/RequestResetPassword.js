import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Form from './styles/Form';
import ErrorMessage from './ErrorMessage';

const REQUEST_RESET_PASSWORD_MUTATION = gql`
  mutation REQUEST_RESET_PASSWORD_MUTATION($email: String!) {
    requestResetPassword(email: $email) {
      message
    }
  }
`;

class RequestResetPassword extends Component {
  state = {
    email: ''
  };

  onChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  };

  onSubmit = async (e, requestResetPasswordMutation) => {
    e.preventDefault();
    await requestResetPasswordMutation();
    this.setState({
      email: ''
    });
  };

  render() {
    return (
      <Mutation
        mutation={REQUEST_RESET_PASSWORD_MUTATION}
        variables={this.state}
      >
        {(requestResetPassword, { error, loading, called }) => (
          <Form
            method="post"
            onSubmit={e => this.onSubmit(e, requestResetPassword)}
          >
            <fieldset disabled={loading} aria-busy={loading}>
              <h2>Reset your password</h2>
              <ErrorMessage error={error} />
              <label htmlFor="email">
                Email
                <input
                  type="email"
                  name="email"
                  placeholder="Email"
                  value={this.state.email}
                  onChange={this.onChange}
                />
              </label>
              <button type="submit">Request!</button>
              {!error && !loading && called && <p>Reset link emailed!</p>}
            </fieldset>
          </Form>
        )}
      </Mutation>
    );
  }
}

export default RequestResetPassword;
export { REQUEST_RESET_PASSWORD_MUTATION };
