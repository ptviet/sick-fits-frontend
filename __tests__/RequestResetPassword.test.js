import { MockedProvider } from 'react-apollo/test-utils';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import wait from 'waait';
import RequestResetPassword, {
  REQUEST_RESET_PASSWORD_MUTATION
} from '../components/RequestResetPassword';

const mocks = [
  {
    // when
    request: {
      query: REQUEST_RESET_PASSWORD_MUTATION,
      variables: { email: 'ptviet@live.com' }
    },
    // then
    result: {
      data: {
        requestResetPassword: {
          message: 'Request sent.',
          __typename: 'Message'
        }
      }
    }
  }
];

describe('<RequestResetPassword />', () => {
  let wrapper;
  it('should render and match the snapshot', () => {
    wrapper = mount(
      <MockedProvider>
        <RequestResetPassword />
      </MockedProvider>
    );

    const form = wrapper.find('form');
    expect(toJson(form)).toMatchSnapshot();
  });

  it('should call the mutation', async () => {
    wrapper = mount(
      <MockedProvider mocks={mocks}>
        <RequestResetPassword />
      </MockedProvider>
    );

    // Simulate typing an email
    wrapper.find('input').simulate('change', {
      target: { name: 'email', value: 'ptviet@live.com' }
    });

    // Submit the form
    wrapper.find('form').simulate('submit');
    await wait();
    wrapper.update();
    expect(wrapper.find('p').text()).toContain('Reset link emailed');
  });
});
