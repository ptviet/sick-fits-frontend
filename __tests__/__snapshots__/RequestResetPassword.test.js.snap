// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestResetPassword /> should render and match the snapshot 1`] = `
<form
  className="Form-sc-1xszr8q-0 fCpNnG"
  method="post"
  onSubmit={[Function]}
>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <h2>
      Reset your password
    </h2>
    <DisplayError
      error={Object {}}
    />
    <label
      htmlFor="email"
    >
      Email
      <input
        name="email"
        onChange={[Function]}
        placeholder="Email"
        type="email"
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Request!
    </button>
  </fieldset>
</form>
`;
